src/Domain/CoordinateMaps/CylindricalEndcap.cpp

```diff
@@ -0,0 +1,211 @@
+  //    of parameters where sphere_one and sphere_two are disjoint;
+  //    allowing those parameter choices would involve much more
+  //    complicated logic to determine whether the map produces shapes
+  //    with sharp angles or large jacobians, and it would involve more
+  //    complicated unit tests to cover those possibilities.
+
+  // First test for invertibility.
+
+  // Consider the intersection of sphere_one and the plane formed by
+  // z_plane.  Call it circle_one.  Consider the cone with apex
+  // center_one that intersects sphere_one on circle_one. This cone
+  // has an opening angle 2*theta.  Call the cone 'cone_one'.
+  const double cos_theta = (z_plane - center_one[2]) / radius_one;
+
+  // Now consider a different cone, cone_new, constructed so that
+  // cone_new and cone_one intersect each other on circle_one at right
+  // angles.  The apex of cone_new is cone_new_apex, defined as
+  // follows:
+  const std::array<double, 3> cone_new_apex = {
+      center_one[0], center_one[1], center_one[2] + radius_one / cos_theta};
+  // Cone_new opens in the -z direction with opening angle 2*(pi/2-theta).
+
+  // A necessary condition for invertibility is that proj_center lies
+  // either inside of cone_new or inside of the reflection of cone_new
+  // (a cone with apex cone_new_apex but opening in the +x direction
+  // with opening angle 2*(pi/2-theta)).
+
+  // Determine the angle of proj_center relative to cone_new_apex. Call this
+  // angle alpha.
+  const double dist_cone_proj = sqrt(square(cone_new_apex[0] - proj_center[0]) +
+                                     square(cone_new_apex[1] - proj_center[1]) +
+                                     square(cone_new_apex[2] - proj_center[2]));
+  const double cos_alpha = (cone_new_apex[2] - proj_center[2]) / dist_cone_proj;
+
+  // Now make sure that alpha < pi/2-theta.
+  // The cone on either side of cone_new_apex is ok, so we use abs below.
+  ASSERT(acos(abs(cos_alpha)) < abs(asin(cos_theta)),
+         "The arguments passed into the CylindricalEndcap constructor "
+         "yield a noninvertible map.");
+
+  // Another necessary condition for invertibility is that proj_center
+  // cannot lie between sphere_one and cone_new_apex.
+  const double proj_radius_one = sqrt(square(center_one[0] - proj_center[0]) +
+                                      square(center_one[1] - proj_center[1]) +
+                                      square(center_one[2] - proj_center[2]));
+  ASSERT(proj_center[2] > cone_new_apex[2] or proj_center[2] < z_plane or
+             proj_radius_one < radius_one,
+         "The arguments passed into the CylindricalEndcap constructor "
+         "yield a noninvertible map.");
+
+  // Other sanity checks that may be relaxed if there is more logic
+  // added and more unit tests to test these cases.
+
+  ASSERT(proj_center[2] < z_plane,
+         "CylindricalEndcap: The map hasn't been tested for this "
```

Comment:
this configuration -> (z component of the projection center) < z_{plane} or something along those lines (spell out the condition)